### PR TITLE
Remove provider guards, change RabbitWorkState to use IMessage, update versions of RabbitMQ projects

### PR DESCRIPTION
**Pull request by tiggerite**
The compression and encryption providers aren’t mandatory for these steps where they are guarded against null. RabbitWorkState should probably be SendMessage/SendMessageSent with IMessage type now. Also updated all the project versions so nothing is out of date (hopefully) and ready to NuGet publish

### CODE REVIEW

File: src/HouseofCat.RabbitMQ.Dataflows/ConsumerDataflow.cs [hunk 1]
```diff
@@ -173,7 +173,6 @@ public ConsumerDataflow<TState> WithReadyToProcessBuffer(int boundedCapacity)
             int? boundedCapacity = null)
         {
             Guard.AgainstNull(suppliedStep, nameof(suppliedStep));
-            Guard.AgainstNull(_encryptionProvider, nameof(_encryptionProvider));
```
Comment (tristan-hyams) [inline src/HouseofCat.RabbitMQ.Dataflows/ConsumerDataflow.cs:176]: Good fix.

**Comment by tristan-hyams** [comment]
Added prod support fix for a bug that landed on my desk.

Utf8Json passed current requirements for File deserialization. Dictionary<string, object>() is now supported for Args both in Topologer and Consumer Topology Auto Creation.

Related PR in legacy library still in use.
https://github.com/houseofcat/RabbitMQ.Core/commit/88a4e86135b452391d64b5c6da7aaecddd40b3e3#diff-6fea380fae78a12de3941e66af7d2111eb85eb8e43ac991df7a14ca9907148aa

**Comment by tristan-hyams** [comment]
I avoided interfaces for properties a longtime ago due to various issues with deserialization.

Currently working on solves via Integration Tests which needed updates.
`NotSupportedException: Deserialization of interface types is not supported. Type 'HouseofCat.RabbitMQ.IMetadata'.`

**Review by tiggerite — COMMENTED** [review]
I actually changed the property originally to MessageId in my first PR but went safe with the function, this is much cleaner.

File: src/HouseofCat.RabbitMQ/Messages/Letter.cs [hunk 2]
```diff
@@ -95,7 +95,7 @@ public Letter(string exchange, string routingKey, byte[] data, string id, byte p
         public Letter Clone()
         {
             var clone = this.Clone<Letter>();
-            clone.LetterMetadata = LetterMetadata.Clone<LetterMetadata>();
+            clone.LetterMetadata = (LetterMetadata)LetterMetadata.Clone<LetterMetadata>();
```
Comment (tiggerite) [inline src/HouseofCat.RabbitMQ/Messages/Letter.cs]: We could change the extension method too for cleaner code:

`public static T Clone<T>(this IMetadata metadata) where T: IMetadata, new() `

Comment (tiggerite) [reply src/HouseofCat.RabbitMQ/Messages/Letter.cs:98]: Oh the markup removed the T generic around the Clone, but you get the gist hopefully..

Comment (tristan-hyams) [reply src/HouseofCat.RabbitMQ/Messages/Letter.cs]: Good point, keeping with the current trend.

**Comment by tristan-hyams** [comment]
Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.

Updating RabbitMQ server / Erlang locally since they are out of date.

**Comment by tiggerite** [comment]
> Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> 
> Updating RabbitMQ server / Erlang locally since they are out of date.



> Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> 
> Updating RabbitMQ server / Erlang locally since they are out of date.

Hmm none of them rely on Clone of the metadata do they? Because with that being new T now its probably pretty slow, on a par with Activator.CreateInstance. Could be swapped out for your Reflection library New version actually as it’s a default constructor.

**Comment by tristan-hyams** [comment]
> > Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> > Updating RabbitMQ server / Erlang locally since they are out of date.
> 
> > Still fixing unit tests. Getting unusual behavior on Publish. I am losing a few publishes.
> > Updating RabbitMQ server / Erlang locally since they are out of date.
> 
> Hmm none of them rely on Clone of the metadata do they? Because with that being new T now its probably pretty slow, on a par with Activator.CreateInstance. Could be swapped out for your Reflection library New version actually as it’s a default constructor.

Already did :) Still about 5 times faster than new T();... so sadly (I prefer code readability than reflection hackery) still necessary for hot loops.

Regarding Publish, this looks like its been a hidden issue for a while now. It ties into RabbitMQ.Client v6.2.1. I slow down publishing it doesn't get lost.

File: src/HouseofCat.RabbitMQ/Messages/PublishReceipt.cs [hunk 3]
```diff
@@ -3,18 +3,17 @@ namespace HouseofCat.RabbitMQ
     public interface IPublishReceipt
     {
         bool IsError { get; set; }
+        public string MessageId { get; set; }
```
Comment (tiggerite) [inline src/HouseofCat.RabbitMQ/Messages/PublishReceipt.cs]: Oops the public probably shouldn’t be on an interface right?

Comment (tristan-hyams) [reply src/HouseofCat.RabbitMQ/Messages/PublishReceipt.cs]: No, its a mistake I will fix.

Publishing bug had me frazzled.

**Comment by tristan-hyams** [comment]
Important Pub/Sub test failed due to XUnit2 Test ThreadPool MaxConcurrency getting tripped from the library and unable to resolve itself as the main Test invocation thread was stuck in While ( await Task.Delay(1) ) ; Alternatively, I could have made the delay longer than 1 and the issue successfully resolved itself.

Adding Task.Yield() in the test method before the loops fixed test and the lost publishes. Prior, the test never finished and would time out. Debugging the behavior left me confused because it never finished publishing which has never happened before.

Suffice to say, be careful with high performant integration tests... they are not benchmarks don't treat them that way.

All Integration tests - which I need more of - are all working as expected again. They had gotten a little outdated prior to recent changes, but still good to go back and fix them.

Note to self: Definitely room for improvement of verifying publishing is happening -> thus easier to test alongside consumers.

**Comment by tiggerite** [comment]
> Adding Task.Yield() in the test method before the loops fixed test and the lost publishes.

That’s a relief, I thought HouseofCat RabbitMQ Client was going to make a comeback ;)

**Pull request merged**